### PR TITLE
Implementing barbuc

### PR DESCRIPTION
**Pull request by baudtack**
Added |$ rune as a mold builder

### CODE REVIEW

File: pkg/arvo/sys/hoon.hoon [hunk 1]
```diff
@@ -6654,6 +6669,7 @@
     {$yell p/(list hoon)}                               ::  render as tank
     {$xray p/manx:hoot}                                 ::  ;foo; templating
   ::                                            ::::::  cores
+    [%brbs sample=spec body=spec]
```
Comment (philipcmonk) [inline pkg/arvo/sys/hoon.hoon]: Please use the column 57 `::` and old-style `{}/` for consistency with surrounding code.  I'd be happy with a separate commit that changed it for all of `+hoon`, but I don't like mixing styles within a single type definition.

File: pkg/arvo/sys/hoon.hoon [hunk 2]
```diff
@@ -16793,7 +16811,27 @@
           ==
         ==
       ==
-    ::
+    ::  parses [a b c] or ==  a  b  c  ==
```
Comment (philipcmonk) [inline pkg/arvo/sys/hoon.hoon]: Why does this having a leading `==`?  There's a precedent for a `==` that isn't at the end of the twig (`=:`), but none for circumfix `==`.

File: pkg/arvo/sys/hoon.hoon [hunk 3]
```diff
@@ -16793,7 +16811,27 @@
...
+    ++  lynx
```
Comment (philipcmonk) [inline pkg/arvo/sys/hoon.hoon]: Hmm, so this is implemented as a parse-time macro?  Does that mean it doesn't even need an entry in `+hoon`?

We usually add synthetic runes by adding an entry to `+open` instead of the parser.  This allows for better composability -- one synthetic rune can desugar to another.

File: pkg/arvo/sys/hoon.hoon [hunk 4]
```diff
@@ -291,22 +292,25 @@
   ::  sample type of `*`.
   $-(* *)
 ::
-+*  list  [item]
+++  list
+  |$  [item]
```
Comment (pilfer-pandex) [inline pkg/arvo/sys/hoon.hoon]: Can one avoid brackets if one has a single argument?

File: pkg/arvo/sys/hoon.hoon (hunk 3 above)
Comment (pilfer-pandex) [reply pkg/arvo/sys/hoon.hoon]: Right. This seems to be mixing parser logic and desugaring logic in one place. Can you please parse this to a %brbs and then desugar that in open:ap?

**Comment by ohAitch** [comment]
Definite agree on not having == act as an open paren, nor requiring extra brackets on the common case with one face: in general a custom parser seems like overkill here, just specify it as a toga and call it a day.

It might even be warranted to have `|$(item=mold ...)` instead of `|$(item ...)` with an implicit `=_*`, leaving the only distinction between `|$` and `|*` that the body is itself in mold form - further rolling back the `+*` "magic"

File: pkg/arvo/sys/hoon.hoon (hunk 1 above)
Comment (baudtack) [reply pkg/arvo/sys/hoon.hoon]: Funny, I remember saying this exact same thing to someone. *stares at @frodwith *

File: pkg/arvo/sys/hoon.hoon (hunk 2 above)
Comment (baudtack) [reply pkg/arvo/sys/hoon.hoon]: Untrue. There is a rune that uses == I just... forget which one it is. I'll get back to you.

File: pkg/arvo/sys/hoon.hoon (hunk 4 above)
Comment (baudtack) [reply pkg/arvo/sys/hoon.hoon]: Not as written, no.

File: pkg/arvo/sys/hoon.hoon (hunk 2 above)
Comment (ohAitch) [reply pkg/arvo/sys/hoon.hoon]: ~% possibly? That's still in a fundamentally close-paren capacity iirc, == meaning both ] and [ is just confusing

Comment (baudtack) [reply pkg/arvo/sys/hoon.hoon]: Yes. ~%, thank you. In what way do those function as close-parens? the first one I mean?

File: pkg/arvo/sys/hoon.hoon (hunk 1 above)
Comment (baudtack) [reply pkg/arvo/sys/hoon.hoon]: Also, yes I'll change this, but only to make you happy, and only because I'm going to be changing all of this Soon (tm) anyway.

**Comment by baudtack** [comment]
The tall form of the names was added to be consistent with the fact that things should generally have both forms to allow for layout flexibility. Someone will want to have a mold builder that has a large number of terms. There is another rune that uses == thing == which is ~%, though @ohAitch is saying that they are both functioning as closing parens.

**Comment by ohAitch** [comment]
It's possible I am mistaken, but that would still point towards also changing `~%` - not sure why its syntax doesn't just use a `:~` tbh, I'm guessing it's a wart dating back to it by necessity being one of the first runes necessary to bootstrap the system.

I think "anyone who wants twenty mold parameters can use a `|*` `$,` directly" is a reasonable position, though an argument can be made for instead allowing the
```
[ a  b  c  d  ]
```
irregular form; `|$` introducing a whole new syntax / parser mode is the main thing I am objecting to here, the syntax being inconsistent with existing conventions in various ways was merely enabled by this.

File: pkg/arvo/sys/hoon.hoon (hunk 2 above)
Comment (frodwith) [reply pkg/arvo/sys/hoon.hoon]: have a look at the ~% on ++ut

File: pkg/arvo/sys/hoon.hoon (hunk 3 above)
Comment (frodwith) [reply pkg/arvo/sys/hoon.hoon]: This is an artifact of having translated from +* (where it makes sense that it's a parser macro). The type should have (list term).

**Comment by joemfb** [comment]
`~%` does in fact use as both prefix and suffix for the hook list. In general, the hints have a variety of strange and old syntactic forms. I don't think we should use them as examples for new runes. I like @philipcmonk's idea if we have to have a tall form `|$`.

File: pkg/arvo/sys/hoon.hoon (hunk 2 above)
Comment (philipcmonk) [reply pkg/arvo/sys/hoon.hoon]: Ah, yes.  It's still weird, but at least there's a precedent.

**Comment by ohAitch** [comment]
Hm, I like tall
```
|$  a  b  c
    d  e  f
  ==
```

in the world where it's the common case, but having both it and wide `|$  [a b c d e f]` is weird(where does the `[` go?), and actually one-per line
```
|$  a
    b
    c
  ==
d
```
seems _much_ better spelled
```
|$  $:  a=mold
        b=mold
        c=mold
    ==
d
```


`|$  [a]` seems okay iff <~60% of `|$`s are in fact single-argument, and otherwise is too in your face about reminding you that you could hypothetically use |$ with multiple names(but probably don't want to)

File: pkg/arvo/sys/hoon.hoon (hunk 2 above)
Comment (pilfer-pandex) [reply pkg/arvo/sys/hoon.hoon]: ~% really isn’t a good precedent. Let’s not make more runes like it.

**Comment by pilfer-pandex** [comment]
Philip’s idea for a tall form seems like the best way forward, if you want to add one.

**Comment by pilfer-pandex** [comment]
Actually, do we even need the ==?

Wouldn’t

    |$  a  b  c  $% ...

just work? One can rely purely on being in the last position to make the syntactic and semantic distinction between the “type parameters” and the “result type”, no?

**Comment by philipcmonk** [comment]
There's a mental (and actual) backtracking to that which we try to avoid.  We try to always know what's coming next.  This gives more precise syntax errors, but I also think it makes things simpler to mentally parse.

**Comment by ohAitch** [comment]
More concretely, we wouldn't want to break

```
++  lone
  |$  a
  a
++  etc
```

**Comment by baudtack** [comment]
@jtobin @philipcmonk @joemfb @pilfer-pandex I believe you will now find your comments addressed.

**Comment by jtobin** [comment]
(Also, don't worry about the pill conflicts -- I can sort those out for you when I merge this.)

**Comment by baudtack** [comment]
> (Also, don't worry about the pill conflicts -- I can sort those out for you when I merge this.)

great! what causes those? Have I done my process wrong? How I've been working is making changes, soliding pills in a fakezod as I go. Then when I'm done, because I don't trust myself, i make a new branch and cherry-pick my commits over one at a time running `make pills` and making a commit for the pills and then when i'm done squashing them into the commit they match with.

**Comment by joemfb** [comment]
@baudtack, the pill conflicts are just intervening changes to master. You could try to stay more up-to-date with master when working, but, if anyone else makes changes that get merged before yours, the conflicts will return. This is the main reason why the current approach is unsustainable.

**Comment by joemfb** [comment]
CI appeared to complete successfully, but hung producing the final report. I've restarted the build for now. We'll need to investigate further if the test harness is subject nondeterministic failures.

**Comment by baudtack** [comment]
> I have several style criticisms of the new parsing rules, but I will (nobly) restrain myself (after all, the hoon parser is generally less than stylistically ideal).

Yeeeah. Honestly, I really kind of hate using == here... it feels... wrong. But I also didn't want to start the "lets spend 50 years bike shedding what the closing rune for this should look like" thing. `==` is passable i suppose. I think your comment w/r/t using braces to call out that this is weird isn't bad, [ and ] are so ubiquitous we'd need to use like < > or something.



> Also, I'm afraid of how @baudtack might react if I request any more changes ... ;)

The advantage of being 700 miles away is you can't hear me throw things though :-p

**Comment by philipcmonk** [comment]
PR approvals are like seed investments, once you get a lead, everyone falls in line...

**Comment by ohAitch** [comment]
Having the tall form seems fine as long as we flag it a style violation to
catch anyone using it!
